feat(linter): Add `linterOptions.typeAware`

connorshea opened the pull request:
Built with help from GitHub Copilot + Raptor mini (GitHub's preview model).

Part of #15972.

This allows setting the `typeAware` option in the config file. It is nested under the new top-level `linterOptions` field. The name was chosen as it's the one used by ESLint and will also hold things like the `reportUnusedDisableDirectives` setting in the future.

The goal of this to allow teams to set this in the config file, and that'll help ensure that everyone on a project is using the same linter rules. Otherwise, when I adopt this at work (hopefully in the near-ish future) I will need to have every dev on the team do unnecessary manual configuration.

Without this, everyone would need to configure their editor settings to enable type-aware lint rules and also ensure that they're running oxlint with the right command if they use it via the CLI.

My hope is that more linter options will be added under `linterOptions` in the near-ish future to enable more consistency like this, but this seemed like one of the more important options to add directly to the config file.

Examples:

```jsonc
{
  "linterOptions": {
    "typeAware": false
  },
  "rules": {
    /// Even though it's set to `error,` this is NOT run because it is a type-aware rule and we have `typeAware: false`.
    "typescript/no-floating-promises": "error"
  }
}
```

```jsonc
{
  "linterOptions": {
    "typeAware": true
  },
  "rules": {
    /// This is a type-aware rule and will be run because `typeAware` is true in linterOptions.
    "typescript/no-floating-promises": "error"
  }
}
```

TODO:

- [x] This should inform whether type-aware rules are enabled in the LSP/VS Code extension. If the user has not set a setting to specifically override the config file.
- [ ] Maybe mention this config option in the `--type-aware` flag docs.
- [x] Get this working in nested configs
- [x] Should `LintConfig` have a `linter_options` field at all, is that what determines the available options for a nested config file? Or should it be excluded like categories are?
- [x] Should we add a `--disable-type-aware` flag to disable the behavior, for the case where the config has `typeAware: true`? Decision: Let's defer this for now.
- [ ] Add more tests? (LSP test added, need to add a test for nested configs as well)

connorshea committed:
feat(linter): Add typeAware config option to the oxlint config file.

This will ensure that teams can standardize whether they want type-aware linting enabled or not via their config file, rather than relying solely on the CLI flag and individual developers enabling type-aware rules in their editors.

Mostly built with GitHub Copilot + Raptor mini.

This still needs work to ensure that the flag works with the VS Code extension, and also to potentially nest this setting inside a `linterOptions` object.

The documentation phrasing should also be improved. And I'm assuming the performance here may need to be improved. Nested configs should also error if they have this value set. And a few more tests should be added.

connorshea committed:
Minor improvements.

connorshea committed:
Nest typeAware under the linterOptions key.

connorshea committed:
Iterating further

connorshea committed:
Updating type_aware to be Option<bool> so we can distinguish between unset, true, and false.

connorshea committed:
Rephrase and regenerate snaps.

connorshea committed:
Add tests to ensure the typeAware setting works correctly in regards to overrides.

connorshea committed:
Add more tests for the behavior of the typeAware config option.

connorshea committed:
Remove outdated todo.

connorshea committed:
Update the option docs for linterOptions/typeAware.

Ensure the intent and behavior are clear.

autofix-ci[bot] committed:
[autofix.ci] apply automated fixes

connorshea committed:
Fix typo.

connorshea committed:
Update snaps.

theoludwig commented:
> My hope is that more linter options will be added under linterOptions in the near-ish future

Good idea. :+1: In addition to `typeAware` and `reportUnusedDisableDirectives`, we could also add `typeCheck` and `denyWarnings`.

connorshea committed:
Update config snap.

connorshea committed:
Get the typeAware option working with the LSP and nested configs.

This should now work properly for nested configs and in editors via the LSP connection.

autofix-ci[bot] committed:
[autofix.ci] apply automated fixes

connorshea committed:
Update vscode readme.

connorshea committed:
Add snapshot tests for linter typeAware option and ensuring it overrides the config file setting.

The snapshots are incorrect currently. I need to figure out why the type-aware rule is not enabled currently with this setup...

connorshea committed:
fmt

connorshea committed:
The snapshots now display the expected result, so this works with the LSP.

I am dumb. Thank you AI for figuring out the problem...

connorshea committed:
fmt

connorshea committed:
Remove unnecessary comment.

connorshea committed:
Update the linterOptions schema slightly.

connorshea committed:
Add snapshot tests for typeAware in nested configs.

Currently it does not seem to work as one file is missing a lint violation when it should have it. But I'm unsure whether that's because of a bug in the code or a misconfiguration of the test. Will maybe need help on that.

This file should have a diagnostic but does not: `crates/oxc_language_server/src/linter/snapshots/fixtures_linter_tsgolint_nested_type_aware@nested__test-with-no-lsp-config.ts.snap`

As an aside, it'd be preferable if I had a good way to add descriptions to the frontmatter of these snapshots to say what they should have in them. And it'd be nice if the snaps pointed back to the original rs test somehow. But this is fine for now, I guess.

Sysix reviewed (commented):
great work so far 👍 
I think we need to create a filter for `oxlint` command which files should be passed to `TsGoLintState`. The same should somehow work with the language server.
I build some month ago `LintIgnoreMatcher`, which resolved the `ignorePatterns` for the linted file, maybe you could find/extend some useful code for it :)

connorshea commented:
Closing in favor of #18047 because that rebase scared me
